### PR TITLE
Create newtype wrappers and remove Vec aliases

### PR DESCRIPTION
**Pull request by djc**
Create newtype wrappers around different usages of payload types to avoid type confusion, and remove `Vec` aliases in favor of a trait-based abstraction as discussed in #895. Alternative to parts of #1242.

### CODE REVIEW

**Review by jsha — APPROVED** [review]
Ooh, I really like this approach a lot. It exposes the `Vec`-ness of these types while also keeping specific types for the items being contained.

The only real comment I have is that TlsListElement, ListLength, and wrapped_payload could use some comments.

File: rustls/src/msgs/codec.rs [hunk 1]
```diff
@@ -205,91 +205,68 @@ impl Codec for u64 {
...
-    let mut ret: Vec<T> = Vec::new();
-    let len = usize::from(u16::read(r)?);
-    let mut sub = r.sub(len)?;
-
-    while sub.any_left() {
-        ret.push(T::read(&mut sub)?);
-    }
-
-    Ok(ret)
+pub trait TlsListElement {
```
Comment (jsha) [inline rustls/src/msgs/codec.rs:264]:
```suggestion
pub(crate) trait TlsListElement {
```

File: rustls/src/msgs/codec.rs [hunk 2]
```diff
@@ -205,91 +205,68 @@ impl Codec for u64 {
...
-    }
-
-    let mut sub = r.sub(len)?;
-
-    while sub.any_left() {
-        ret.push(T::read(&mut sub)?);
-    }
-
-    Ok(ret)
+pub enum ListLength {
```
Comment (jsha) [inline rustls/src/msgs/codec.rs:268]:
```suggestion
pub(crate) enum ListLength {
```

**Review by jsha — CHANGES_REQUESTED** [review]

**Review by cpu — COMMENTED** [review]
This looks great, clever use of a macro :+1:

**Comment by djc** [comment]
Add some comments and privatized the `TlsListElement` trait and `ListLength` type.

**Review by ctz — APPROVED** [review]

**Review by jsha — APPROVED** [review]

**Pull request merged**